From c97b216e112b6e9981d8dd6b176a3c581178aee9 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Thu, 23 Oct 2014 20:01:12 -0400
Subject: [md] dm-bufio: update last_accessed when relinking a buffer

Message-id: <1414094494-22068-3-git-send-email-snitzer@redhat.com>
Patchwork-id: 98336
O-Subject: [RHEL7.2 PATCH 02/24] dm bufio: update last_accessed when relinking a buffer
Bugzilla: 1156161
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1156161

Upstream commit eb76faf53b1ff7a77ce3f78cc98ad392ac70c2a0
Author: Joe Thornber <ejt@redhat.com>
Date:   Tue Sep 30 09:32:46 2014 +0100

    dm bufio: update last_accessed when relinking a buffer

    The 'last_accessed' member of the dm_buffer structure was only set when
    the the buffer was created.  This led to each buffer being discarded
    after dm_bufio_max_age time even if it was used recently.  In practice
    this resulted in all thinp metadata being evicted soon after being read
    -- this is particularly problematic for metadata intensive workloads
    like multithreaded small random IO.

    'last_accessed' is now updated each time the buffer is moved to the head
    of the LRU list, so the buffer is now properly discarded if it was not
    used in dm_bufio_max_age time.

    Signed-off-by: Joe Thornber <ejt@redhat.com>
    Signed-off-by: Mikulas Patocka <mpatocka@redhat.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>
    Cc: stable@vger.kernel.org # v3.2+
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-bufio.c b/drivers/md/dm-bufio.c
index d1ad083..34a4845 100644
--- a/drivers/md/dm-bufio.c
+++ b/drivers/md/dm-bufio.c
@@ -465,6 +465,7 @@ static void __relink_lru(struct dm_buffer *b, int dirty)
  c->n_buffers[dirty]++;
  b->list_mode = dirty;
  list_move(&b->lru_list, &c->lru[dirty]);
+ b->last_accessed = jiffies;
 }
 
 /*----------------------------------------------------------------
-- 
1.7.1